### PR TITLE
SDIT-1957-Update-the-new-index-using-partial-updates-instead-of-an-overall-save

### PR DESCRIPTION
**Pull request by steverendell**
Remove a duplicate RED update on receipt of a domain event

### CODE REVIEW

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt [hunk 1]
```diff
@@ -27,7 +27,7 @@ class IndexListenerService(
       message.additionalInformation.nomsNumber,
       message.additionalInformation.id,
     )
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt:29]: Only need to sync the old(green/blue) index, red is updated selectively

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt [hunk 2]
```diff
@@ -37,13 +37,13 @@ class IndexListenerService(
       message.description,
       message.additionalInformation.prisonerNumber,
     )
-    sync(message.additionalInformation.prisonerNumber, eventType)
+    syncBoth(message.additionalInformation.prisonerNumber, eventType)
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt:40]: Renamed to emphasise that it is updating both old and new red indices

File: hmpps-prisoner-search-indexer/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/listeners/DomainEventListenerIntTest.kt [hunk 3]
```diff
@@ -96,6 +99,12 @@ class DomainEventListenerIntTest : IntegrationTestBase() {
         assertThat(currentIncentive?.level?.code).isEqualTo("STD")
       }
     }
+    verify(prisonerSpyBeanRepository, times(1)).save(any(), eq(SyncIndex.GREEN))
+    verify(prisonerSpyBeanRepository, never()).save(any(), eq(SyncIndex.BLUE))
+    verify(prisonerSpyBeanRepository, never()).save(any(), eq(SyncIndex.RED))
+    verify(prisonerSpyBeanRepository, times(1)).updateIncentive(eq(prisonerNumber), any(), eq(SyncIndex.RED), any())
+    verify(prisonerSpyBeanRepository, never()).updatePrisoner(eq(prisonerNumber), any(), eq(SyncIndex.GREEN), any())
+    verify(prisonerSpyBeanRepository, never()).updatePrisoner(eq(prisonerNumber), any(), eq(SyncIndex.RED), any())
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/listeners/DomainEventListenerIntTest.kt:107]: Nail down exectly what index updates occur

File: hmpps-prisoner-search-indexer/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/listeners/OffenderEventListenerIntTest.kt [hunk 4]
```diff
@@ -33,6 +41,47 @@ class OffenderEventListenerIntTest : IntegrationTestBase() {
...
       assertThat(prisoner?.prisonerNumber).isEqualTo(prisonerNumber)
     }
+
+    verify(prisonerSpyBeanRepository, times(1)).save(any(), eq(SyncIndex.GREEN))
+    verify(prisonerSpyBeanRepository, never()).save(any(), eq(SyncIndex.BLUE))
+    verify(prisonerSpyBeanRepository, times(1)).save(any(), eq(SyncIndex.RED))
+    verify(prisonerSpyBeanRepository, never()).updateIncentive(eq(prisonerNumber), any(), eq(SyncIndex.RED), any())
+    verify(prisonerSpyBeanRepository, never()).updatePrisoner(eq(prisonerNumber), any(), eq(SyncIndex.GREEN), any())
+    verify(prisonerSpyBeanRepository, never()).updatePrisoner(eq(prisonerNumber), any(), eq(SyncIndex.RED), any())
+  }
```
Comment (steverendell) [inline hmpps-prisoner-search-indexer/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/listeners/OffenderEventListenerIntTest.kt:51]: Calls save() for the red index if the prisoner does not exist, but updates if it does exist

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt [hunk 5]
```diff
@@ -126,14 +148,23 @@ class IndexListenerService(
...
+   */
+  private fun syncBoth(prisonerNumber: String, eventType: String): Prisoner? =
     nomisService.getOffender(prisonerNumber)?.run {
-      reindexPrisoner(ob = this, eventType)
+      reindexPrisonerNew(ob = this, eventType)
+      reindexPrisonerOld(ob = this, eventType)
     } ?: null.also { log.warn("Sync requested for prisoner {} not found", prisonerNumber) }
 
-  private fun sync(bookingId: Long, eventType: String): Prisoner? =
+  private fun syncOld(prisonerNumber: String, eventType: String): Prisoner? =
```
Comment (petergphillips) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt:160]: might be clearer if this was `syncGreenBlue` and the other was `syncRed` for now

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt [hunk 6]
```diff
@@ -114,6 +114,28 @@ class IndexListenerService(
...
 
+  private fun reindexPrisonerOld(ob: OffenderBooking, eventType: String): Prisoner? =
+    indexStatusService.getIndexStatus()
+      .run {
+        if (activeIndexesEmpty()) {
+          log.info("Ignoring update (old) of prisoner {} as no indexes were active", ob.offenderNo)
+          null
+        } else {
+          val prisoner = prisonerSynchroniserService.reindex(ob, activeIndexes(), eventType)
+          prisoner
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt:125]: No val is needed here

**Review by petergphillips — APPROVED** [review]
LGTM

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt (hunk 6 above)
Comment (mikehalmamoj) [reply hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt:125]: There's another one on line 111 which might be clearer as:
```
prisonerSynchroniserService.reindex(ob, activeIndexes(), eventType)
  .also { prisonerSynchroniserService.reindexUpdate(ob, eventType) }
```

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt (hunk 5 above)
Comment (mikehalmamoj) [reply hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt:160]: Same applies to the `reindexPrisoner...` methods.

Comment (steverendell) [reply hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt]: yes good idea

File: hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt (hunk 6 above)
Comment (steverendell) [reply hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/IndexListenerService.kt:125]: I could just reorder them

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**